src/debug_sdfs_ed64.c

```diff
@@ -0,0 +1,598 @@
+	if (ret_val == RES_OK) return RES_OK;
+
+	// Do error cleanup
+	everdrive_sd_change_mode(ED64_SD_MODE_NONE, 0);
+
+	return ret_val;
+}
+
+static DRESULT fat_disk_write_everdrive(const BYTE* buff, LBA_t sector, UINT count) {
+	_Static_assert(FF_MIN_SS == 512, "this function assumes sector size == 512");
+	_Static_assert(FF_MAX_SS == 512, "this function assumes sector size == 512");
+
+	uint8_t result;
+	DRESULT ret_val = RES_OK;
+
+	// Overclock the PI
+	uint32_t old_pw = PI_regs->dom1_pulse_width;
+	io_write((uint32_t)&PI_regs->dom1_pulse_width, 0x09);
+
+	if(!everdrive_sd_change_mode(ED64_SD_MODE_BLOCK_WRITE, sector)) {
+		ret_val = RES_ERROR; goto cleanup;
+	};
+
+	uint16_t crc[4], timeout;
+
+	for (int i=0;i<count;i++)
+	{
+		set_everdrive_sd_bitlen(2);
+		everdrive_sd_write_data(0xff);
+		everdrive_sd_write_data(0xf0); // Pull all lines low to start transfer
+
+		set_everdrive_sd_bitlen(4);
+
+		if (((uint32_t)buff & 7) == 0)
+		{
+			data_cache_hit_writeback(buff, 512);
+			dma_write_raw_async(buff, ED64_BASE_ADDRESS + ED64_SD_IO_BUFFER, 512);
+		}
+		else
+		{
+			typedef uint32_t u_uint32_t __attribute__((aligned(1)));
+
+			uint32_t* dst = (uint32_t*)(ED64_BASE_ADDRESS + ED64_SD_IO_BUFFER);
+			u_uint32_t* src = (u_uint32_t*)buff;
+			for (int i = 0; i < 512/16; i++)
+			{
+				uint32_t a = *src++; uint32_t b = *src++; uint32_t c = *src++; uint32_t d = *src++;
+				*dst++ = a;          *dst++ = b;          *dst++ = c;          *dst++ = d;
+			}
+		}
+
+		everdrive_sd_crc16((void*)buff, crc);
+
+		data_cache_hit_writeback(crc, 8);
+		dma_write(crc, ED64_BASE_ADDRESS + ED64_SD_IO_BUFFER, 8);
+
+		// Each read will shift 4 bit of parallel data. dat0 will go low when we
+		// have the data response token's status. Read it from the same line
+		// once found. Swiching to command mode does not work here although they
+		// should be using the same line for the response?
+		set_everdrive_sd_bitlen(1);
+		timeout = 1024;
+		do {
+			if (!timeout--) {
+				debugf("Write resp timeout\n");
+				ret_val = RES_ERROR; goto cleanup;
+			}
+			result = everdrive_sd_read_data();
+		} while (result != 0xFE);
+
+		result = (everdrive_sd_read_data() & 1) << 2;
+		result |= (everdrive_sd_read_data() & 1) << 1;
+		result |= (everdrive_sd_read_data() & 1);
+
+		if (result == 0b101) {
+			debugf("Write CRC mismatch\n");
+			ret_val = RES_ERROR; goto cleanup;
+		}
+
+		if (result != 0b010) {
+			debugf("Write Error\n");
+			ret_val = RES_ERROR; goto cleanup;
+		}
+
+		// Consume all remaining data
+		timeout = -1;
+		while(everdrive_sd_read_data() != 0xFF) {
+			if(!timeout--) {
+				debugf("Flush data timeout\n");
+				ret_val = RES_ERROR; goto cleanup;
+			}
+		};
+		buff += 512;
+	}
+
+everdrive_sd_address = sector + count;
+
+cleanup:
+	io_write((uint32_t)&PI_regs->dom1_pulse_width, old_pw);
+
+	if (ret_val == RES_OK) return RES_OK;
+
+	// Do error cleanup
+	everdrive_sd_change_mode(ED64_SD_MODE_NONE, 0);
+
+	return ret_val;
```

Comment:
Same here

Reply:
Added a comment to be a little more clearer.